src/main/java/uk/ac/ic/wlgitbridge/bridge/gc/GcJobImpl.java

```diff
@@ -0,0 +1,141 @@
+package uk.ac.ic.wlgitbridge.bridge.gc;
+
+import uk.ac.ic.wlgitbridge.bridge.lock.LockGuard;
+import uk.ac.ic.wlgitbridge.bridge.lock.ProjectLock;
+import uk.ac.ic.wlgitbridge.bridge.repo.ProjectRepo;
+import uk.ac.ic.wlgitbridge.bridge.repo.RepoStore;
+import uk.ac.ic.wlgitbridge.util.Log;
+import uk.ac.ic.wlgitbridge.util.TimerUtils;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Implementation of {@link GcJob} using its own Timer and a synchronized
+ * queue.
+ */
+public class GcJobImpl implements GcJob {
+
+    private final RepoStore repoStore;
+    private final ProjectLock locks;
+
+    private final long intervalMs;
+    private final Timer timer;
+
+    private final Set<String> gcQueue;
+
+    /**
+     * Hooks in case they are needed, e.g. for testing.
+     */
+    private AtomicReference<Runnable> preGc;
+    private AtomicReference<Runnable> postGc;
+
+    /* We need to iterate over and empty it after every run */
+    private final Lock jobWaitersLock;
+    private final List<CompletableFuture<Void>> jobWaiters;
+
+    public GcJobImpl(RepoStore repoStore, ProjectLock locks, long intervalMs) {
+        this.repoStore = repoStore;
+        this.locks = locks;
+        this.intervalMs = intervalMs;
+        timer = new Timer();
+        gcQueue = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        preGc = new AtomicReference<>(() -> {});
+        postGc = new AtomicReference<>(() -> {});
+        jobWaitersLock = new ReentrantLock();
+        jobWaiters = new ArrayList<>();
+    }
+
+    public GcJobImpl(RepoStore repoStore, ProjectLock locks) {
+        this(
+                repoStore,
+                locks,
+                TimeUnit.MILLISECONDS.convert(1, TimeUnit.HOURS)
+        );
+    }
+
+    @Override
+    public void start() {
+        Log.info("Starting GC job to run every [{}] ms", intervalMs);
+        timer.scheduleAtFixedRate(
+                TimerUtils.makeTimerTask(this::doGC),
+                intervalMs,
```

Comment:
That's right. The reason we don't do it initially is that the queue of projects to GC is empty. We've taken the approach that we only queue projects to GC that have been touched recently. So even if the job ran, nothing would happen!